File: Lock_BFS.py
from typing import List


class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        deadends = [int(item) for item in deadends]
        target = int(target)
        init = 0
        queue = [init]
        n_turn = 0
        visited = [0] * 10000
        visited[0] = 1
        if init in deadends:
            return -1
        if target == init:
            return 0
        while len(queue) != 0:
            n_turn += 1
            queue_size = len(queue)
            for k in range(queue_size):
                current = queue.pop(0)
                if current in deadends:
                    continue
                for i in range(4):
                    position_value = int((current / pow(10, i)) % 10)
                    tmp = int(current - position_value * pow(10, i))
                    for j in [-1, 1]:
                        child = int(((position_value + j) % 10) * pow(10, i) + tmp)
                        if child == target:
                            return n_turn
                        if (not child in deadends) and visited[child] == 0:
                            queue.append(child)
                            visited[child] = 1

        return -1


if __name__ == "__main__":
    deadends = ["0201", "0101", "0102", "1212", "2002"]
    target = "0202"
    solution = Solution()
    n_turn = solution.openLock(deadends, target)
    print(n_turn),

File: Lock_BFS.cpp
//
// Created by Golfer on 2020/7/19.
//

#include <cstdio>
#include <queue>
#include <string>
#include <set>

using namespace std;

class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        // change vector to set for quick query
        set<string> deads;
        for (string &item : deadends) {
            deads.insert(item);
        }
        string init = "0000";
        if (deads.find(init) != deads.end()) {
            return -1;
        }
        if (target == init) {
            return 0;
        }
        // visited
        set<string> visited;
        visited.insert(init);
        queue<string> q;
        q.push(init);
        int n_turns = 0;
        while (! q.empty()) {
            ++n_turns;
            int q_size = q.size();
            for (int i = 0; i < q_size; i ++) {
                string current = q.front();
                q.pop();
                for (int j = 0; j < 4; j++) {
                    string child_plus_one = get_child_plus_one(current, j);
                    if (child_plus_one == target) {
                        return n_turns;
                    }
                    if (deads.find(child_plus_one) == deads.end() && visited.find(child_plus_one) == visited.end()) {
                        q.push(child_plus_one);
                        visited.insert(child_plus_one);
                    }
                    string child_minus_one = get_child_minus_one(current, j);
                    if (child_minus_one == target) {
                        return n_turns;
                    }
                    if (deads.find(child_minus_one) == deads.end() && visited.find(child_minus_one) == visited.end()) {
                        q.push(child_minus_one);
                        visited.insert(child_minus_one);
                    }
                }
            }
        }
        return -1;
    }

    string get_child_plus_one(string current, int position) {
        if (current[position] == '9') {
            current[position] = '0';
        }
        else {
            current[position] += 1;
        }
        return current;
    }

    string get_child_minus_one(string current, int position) {
        if (current[position] == '0') {
            current[position] = '9';
        }
        else {
            current[position] -= 1;
        }
        return current;
    }
};


int main() {

    vector<string> deadends = {"0201", "0101", "0102", "1212", "2002"};
    string target = "0202";
    Solution solution = Solution();
    int n_turns = solution.openLock(deadends, target);
    printf("%d", n_turns);
}

